### ISSUE TITLE
ConvertFrom-Json breaking change:  automagically detects timestamps and deserializes to DateTime

### ISSUE THREAD

**Issue by billwolckenlmco** [PowerShell/PowerShell#3378]
ConvertFrom-Json in prior releases of PowerShell left strings that looked like date/timestamps in JSON data alone and deserialized them as strings.  In v6.0.0-alpha.17 on Mac, they are being converted to DateTimes.  This a breaking change, and if it is an intended feature, it should be opt-in via a cmdlet parameter of ConvertFrom-Json.  (In my scenario, I do not want the automagic conversion, since I want the values preserved round-trip, which would be the case if they stayed as strings.  DateTime serialization in ConvertTo-Json assumes my current timezone and is changing the value, plus it formats the zone as "-05:00" unlike the original JSON's "-0400" format and as such the final result is rejected by the Java application it gets passed to.)

Here is my demonstration script:

    $jsonOriginal = '{"info":"2017-01-02T03:04:05.678-0400"}'
    $obj = ConvertFrom-Json $jsonOriginal
    write-host ("Type in object:  " + $obj.info.GetType().FullName)
    $jsonFinal = ConvertTo-Json -compress $obj
    $matches = $jsonOriginal -eq $jsonFinal
    write-host "Round trip preserved:  $matches"
    if (!$matches) {
    	write-host "Original:  $jsonOriginal"
    	write-host "Final:  $jsonFinal"
    }
    write-host "Version:"
    $PSVersionTable

On v6.0.0-alpha.17 on Mac the output is:

	Type in object:  System.DateTime
	Round trip preserved:  False
	Original:  {"info":"2017-01-02T03:04:05.678-0400"}
	Final:  {"info":"2017-01-02T02:04:05.678-05:00"}
	Version:

	Name                           Value
	----                           -----
	PSVersion                      6.0.0-alpha
	PSEdition                      Core
	PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
	BuildVersion                   3.0.0.0
	GitCommitId                    v6.0.0-alpha.17
	CLRVersion
	WSManStackVersion              3.0
	PSRemotingProtocolVersion      2.3
	SerializationVersion           1.1.0.1

On PowerShell v5 on Windows 7, the output is:

	Type in object:  System.String
	Round trip preserved:  True
	Version:
	 
	Name                           Value
	----                           -----
	PSVersion                      5.0.10586.117
	PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
	BuildVersion                   10.0.10586.117
	CLRVersion                     4.0.30319.42000
	WSManStackVersion              3.0
	PSRemotingProtocolVersion      2.3
	SerializationVersion           1.1.0.1

On PowerShell v4, the output is:

	Type in object:  System.String
	Round trip preserved:  True
	Version:
	 
	Name                           Value
	----                           -----
	PSVersion                      4.0
	WSManStackVersion              3.0
	SerializationVersion           1.1.0.1
	CLRVersion                     4.0.30319.42000
	BuildVersion                   6.3.9600.17400
	PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0}
	PSRemotingProtocolVersion      2.2

**Comment by SteveL-MSFT** [comment]
Windows PowerShell used .Net's JSON serializer.  Since that is not in coreFx, for PowerShell Core, we use Newtonsoft.Json which handles thing differently.  In the case of datetime, looking at the Newtonsoft code, they always convert to localtime although they can preserve the UTC.  However, they follow ISO 8601 datetime format which supports both a colon and no colon for the offset, but Newtonsoft decided to [always add the colon](https://github.com/JamesNK/Newtonsoft.Json/blob/aa5f28c09732db5fb3e316433547ea7cdecd8804/Src/Newtonsoft.Json/Utilities/DateTimeUtils.cs#L757).  I would suggest creating an issue in their repo to support a new DateTimeZoneHandling value so that at least the colon isn't inserted and then we can expose a parameter to support that.

**Comment by rmbolger** [comment]
I realize this is a long closed issue at this point. But was this breaking change documented anywhere? I'm having trouble finding any references to it other than this issue. I'd expect it would be somewhere in the what's new, breaking changes, or known issues with PowerShell 6 docs.

**Comment by SteveL-MSFT** [comment]
The JSON cmdlets were re-written for .NET Core since .NET Core didn't have the .NET Framework JSON serializer and NewtonSoft.JSON isn't completely compatible with .NET Framework JSON serializer.  Not sure if this is documented, but I would suggest opening an issue in PowerShell-Docs repo if you don't find sufficient documentation on this.

**Comment by jazzdelightsme** [comment]
The fact that `ConvertFrom-Json` munges my input data *astounds* me (in a bad way).

Now that .NET Core *does* have a good JSON serializer, can we please switch to it?

(Can this Issue be reopened?)

**Comment by vexx32** [comment]
#11198 is currently in progress and I believe we have an existing issue tracking that as well.

@iSazonov should we keep this open to track this specific issue and ensure it gets resolved by those changes?

**Comment by iSazonov** [comment]
For ConvertTo-Json PowerShell Committee approved breaking changes caused by moving to new .Net Json API.
After getting the new ConvertTo-Json we will migrate ConverFrom-Json. Obviously we will follow .Net Json API with many breaking changes. So it makes sense to track issues for old API.

If you think that timestamp deserialization is important you could:
1. Check that is behavior of new .Net Json API
2. If you see it is a wrong you should open an issue in .Net Runtime repo and discuss there while .Net 5.0 under active development.

**Comment by sba923** [comment]
IMVHO this breaking change is.... broken 😜

See https://github.com/PowerShell/PowerShell/issues/13592

**Comment by beatcracker** [comment]
My 2￠- I found about this change when I've tried to write a unit test for a JSON file. I wanted to check if property containing time conforms to specific format (ISO 8601/RFC 3339). I was using `[datetime]::TryParseExact` and couldn't figure out why it fails until I looked into the debugger and saw that the property was converted to `[dateime]` instead of being a plain string.

I get how this could benefit round trip scenarios, but I wish that `ConvertFrom-Json` had an option to disable this behavior. Perhaps a `-Raw` switch?
